include/gridtools/stencil_composition/backend_cuda/launch_kernel.hpp

```diff
@@ -0,0 +1,244 @@
+                GT_CUDA_CHECK(
+                    cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, shared_memory_size));
+                kernel<<<blocks, threads, shared_memory_size>>>(std::move(args)...);
+#ifndef NDEBUG
+                GT_CUDA_CHECK(cudaDeviceSynchronize());
+#else
+                GT_CUDA_CHECK(cudaGetLastError());
+#endif
+            }
+
+            template <class Extent,
+                int_t BlockSizeI,
+                int_t BlockSizeJ,
+                class Fun,
+                std::enable_if_t<!is_empty_ij_extents<Extent>(), int> = 0>
+            GT_FORCE_INLINE void launch_kernel(
+                int_t i_size, int_t j_size, uint_t zblocks, Fun fun, size_t shared_memory_size = 0) {
+                GT_STATIC_ASSERT(is_extent<Extent>::value, GT_INTERNAL_ERROR);
+                GT_STATIC_ASSERT(Extent::iminus::value <= 0, GT_INTERNAL_ERROR);
+                GT_STATIC_ASSERT(Extent::iplus::value >= 0, GT_INTERNAL_ERROR);
+
+                GT_STATIC_ASSERT(std::is_trivially_copyable<Fun>::value, GT_INTERNAL_ERROR);
+
+                static constexpr auto halo_lines = Extent::jplus::value - Extent::jminus::value +
+                                                   (Extent::iminus::value < 0 ? 1 : 0) +
+                                                   (Extent::iplus::value > 0 ? 1 : 0);
+                static constexpr size_t num_threads = BlockSizeI * (BlockSizeJ + halo_lines);
+
+                uint_t xblocks = (i_size + BlockSizeI - 1) / BlockSizeI;
+                uint_t yblocks = (j_size + BlockSizeJ - 1) / BlockSizeJ;
+
+                dim3 blocks = {xblocks, yblocks, zblocks};
+                dim3 threads = {BlockSizeI, BlockSizeJ + halo_lines, 1};
+
+                launch(blocks,
+                    threads,
+                    shared_memory_size,
+                    wrapper<num_threads, BlockSizeI, BlockSizeJ, Extent, Fun>,
+                    std::move(fun),
+                    i_size,
+                    j_size);
+            }
+
+            template <class Extent,
+                int_t BlockSizeI,
+                int_t BlockSizeJ,
+                class Fun,
+                std::enable_if_t<is_empty_ij_extents<Extent>(), int> = 0>
+            GT_FORCE_INLINE void launch_kernel(
+                int_t i_size, int_t j_size, uint_t zblocks, Fun fun, size_t shared_memory_size = 0) {
+
+                GT_STATIC_ASSERT(std::is_trivially_copyable<Fun>::value, GT_INTERNAL_ERROR);
+
+                static const size_t num_threads = BlockSizeI * BlockSizeJ;
+
+                uint_t xblocks = (i_size + BlockSizeI - 1) / BlockSizeI;
+                uint_t yblocks = (j_size + BlockSizeJ - 1) / BlockSizeJ;
+
+                dim3 blocks = {xblocks, yblocks, zblocks};
+                dim3 threads = {BlockSizeI, BlockSizeJ, 1};
+
+                // TODO(anstaf): to investigate if this optimization has impact
```

Comment:
Probably do this TODO before the merge? I would guess that there’s no significant impact and we can save some lines of code here, but who knows…

Reply:
done